### DISCUSSION TITLE
What is the right way to add and display an element?

### DISCUSSION THREAD

**Discussion by reypm** [syntax-tree/unist#104] [Q&A] [answered]
I have a custom plugin where I am getting a node and I would like to add an `html` element. Here is a snippet of the function receiving the node:
```tsx
const processPageInclude = (node: any, productId?: string) => {
    const data = node.data || (node.data = {})
    data.hProperties = node.attributes;

    node.children.unshift({
        type: 'containerDirective',
        children: [{
            type: 'html',
            value: '<div>This is some HTML</div>div>'
        }]
    });

   console.log(node)
}
```
`node` will contain the HTML `This is some HTML` but it never gets rendered to the UI ... what I am missing?

**Comment by ChristianMurphy** [comment]
Welcome @reypm!

A few thoughts.
1. since you are working with HTML here, you may want to create a rehype (HTML) plugin rather than a remark (Markdown) one.
2. If you do decide to do a markdown plugin anyway, the rest of your pipeline matters, you need to configure `remark-rehype` to preserve any raw HTML https://github.com/remarkjs/remark-rehype#optionsallowdangeroushtml (the exact details will look different depending on which of the many ways to use remark/mdx/react-markdown you are using)

**Reply by reypm** [reply]
Hi @ChristianMurphy and thank you for your reply. I am grabbing the data from an external service and it has a markdown on it so I need to process that markdown and return the processed content back to the frontend. 

Here is what my configuration looks like:
```tsx
<ReactMarkdown
    remarkPlugins={[
        remarkGfm,
        [remarkFrontmatter, ['yaml', 'toml']],
        remarkHtml,
        remarkParse,
        remarkBreaks,
        [remarkMermaid],
        remarkDirective,
        [shortcodePlugin, [metaData?.id]] // this is my custom plugin
    ]}
    rehypePlugins={[
        rehypeExternalLinks,
        rehypeSlug,
        rehypeStringify,
        rehipeRaw
    ]}
    remarkRehypeOptions={{allowDangerousHtml: true}}
>
    {data}
</ReactMarkdown>
```
what else could I be missing?

**Reply by ChristianMurphy** [reply]
@reypm could you share a runnable example of the issue you are seeing?
Using https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-create-react-app or https://codesandbox.io/embed/github/remarkjs/.github/tree/main/sandbox-templates/react-markdown-with-create-react-app as a base?

**Reply by reypm** [reply]
@ChristianMurphy Sure, here is my example https://codesandbox.io/s/sleepy-ben-cq2msk ... a few thoughts:
```tsx
  newNode.children.unshift({
    type: "containerDirective",
    children: [
      {
        type: "html",
        value: "<div>reynier</div>"
      },
      {
        type: "html",
        //value: "<div>reynier</div>"
        value: <Markdown>{markdownSource}</Markdown>
      }
    ]
  });
```
- the first children (the one containing `<div>reynier</div>`) alone do work, it does not for me in my local environment so certainly the issue seems to be on my side.
- now, I am trying to use `react-markdown` to parse some markdown content and add it back but that is triggering an error

Parse markdown is part of this plugin functionality since I am reading several custom shortcodes and need to parse them, `my-include{id=some-id}` will grab some JSON data containing markdown from external sources and my idea is to parse such markdown and write back to the page using HTML.

what else I am missing here?

**Reply by ChristianMurphy** [reply]
> it does not for me in my local environment so certainly the issue seems to be on my side.

I'd check dependency versions to make sure they are up-to-date, and would check the build tool configuration to see if it could be causing compilation issues.

> now, I am trying to use `react-markdown` to parse some markdown content and add it back but that is triggering an error

`react-markdown` produces JSX functions, `remarkPlugins` are expected to return `mdast` as JSON.
Please read https://unifiedjs.com/learn/ for a better understanding of the different ASTs.

> https://codesandbox.io/s/sleepy-ben-cq2msk

A few thoughts:
1. `remark-parse` is already included by `react-markdown` duplicating as a direct plugin it could cause conflicts
2. `rehype-stringify` is an alternative way to render, separate from `react-markdown` avoid mixing these two.
3. `remark-html` is another alternative way to render, separate from `react-markdown` avoid mixing these two.
4. `rehype-raw` is mostly about handling raw HTML in the markdown, I'm not seeing any in your example, instead create a rehype plugin to add HTML content

I'd recommend something like the following

<details>
<summary><code>App.tsx</code></summary>

```tsx
// ⚠️ Important! Please make sure the dependencies are up to date.

import React from "react";
import Markdown from "react-markdown";
import remarkDirective from "remark-directive";
import rehypeExternalLinks from "rehype-external-links";
import rehypeSlug from "rehype-slug";
import remarkBreaks from "remark-breaks";
import remarkFrontmatter from "remark-frontmatter";
import remarkGfm from "remark-gfm";
import { visit } from "unist-util-visit";
import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkRehype from "remark-rehype";
import { Root as MdastRoot } from "mdast";
import { Root as HastRoot, ElementContent } from "hast";

const markdownSource = `
# heading

* list
* items

\`\`\`js
function () {
  return "this is a function 2";
}
\`\`\`

here goes some text
:::::custom-shortcode{id="some-link"}
Content
:::::
:::::my-include{id=some-id}
:::::
`;

/**
 * Gives directives a tag name, so they will transfer into the HTML AST (hast)
 */
function remarkDirectiveCustomize() {
  return (tree: MdastRoot) =>
    visit(tree, "containerDirective", (node) => {
      node.data ??= {};
      node.data.hName = "container-directive";
      node.data.hProperties = { type: node.name };
    });
}

/**
 * Find the directive HTML tags and replace them with markdown content
 */
function rehypeRenderDirective() {
  return (tree: HastRoot) =>
    visit(tree, { type: "element", tagName: "container-directive" }, (node) => {
      node.tagName = "div";
      node.properties = {};
      node.children = unified()
        .use(remarkRehype)
        // ⚠️ IMPORTANT! if content is untrusted also santize it with `rehype-sanitize`
        .runSync(unified().use(remarkParse).parse("# example"))
        .children as ElementContent[];
    });
}

const App = () => {
  return (
    <Markdown
      remarkPlugins={[
        remarkGfm,
        remarkDirective,
        [remarkFrontmatter, ["yaml", "toml"]],
        remarkBreaks,
        remarkDirectiveCustomize
      ]}
      rehypePlugins={[rehypeRenderDirective, rehypeExternalLinks, rehypeSlug]}
    >
      {markdownSource}
    </Markdown>
  );
};

export default App;
```

</details>

runnable example: https://codesandbox.io/s/nifty-glade-xlwkry?file=/src/app.tsx:277-2104

**Reply by reypm** [reply]
hey @ChristianMurphy great post! just one thing, how can I use this from within my custom remark plugin `myInclude`? There is where I am doing all the data processing. As you can see in my example `myPlugin` has several cases which depending on the custom shortcode will execute some code each time

**Reply by ChristianMurphy** [reply]
@reypm note in my example in `remarkDirectiveCustomize`
```ts
      node.data.hProperties = { type: node.name };
```

This allows you to pass attributes, such as name (shown above), or any properties attached to the directive over into `hast`.
On the `rehypeRenderDirective` side, you can then switch on those properties, or use those attributes to lookup content.

**Reply by reypm** [reply]
does this mean I should add my switch statement inside that?

**Reply by ChristianMurphy** [reply]
Inside `rehypeRenderDirective`, yes.

**Reply by reypm** [reply]
hi, @ChristianMurphy I am still struggling to get how the data I am pulling from my custom plugin can be passed down to the `rehypeRenderDirective` in this line:

> .runSync(unified().use(remarkParse).parse("# example"))

Currently, it prints out just: `Example` but I want it to print whatever I am gathering from myCustomPlugin, is there a way to achieve that?

**Reply by ChristianMurphy** [reply]
Instead of passing `"# example"` into `unified`, pass the string/text of markdown you are gathering.

**Reply by reypm** [reply]
I must be missing something trivial here ... in the function `remarkDirectiveCustomize` you added the `node.name` as a property but for some reason, I cannot access it inside of `rehypeRenderDirective` which I would need to be able to execute the switch statement (I am open to suggestions on this one since `name` is not accessible):

```
switch (node.name) {
  case "my-include":
    myCustomPlugin(node);
    break;
  case "my-include-2":
    myCustomPlugin(node);
    break;
  case "my-include-3":
    myCustomPlugin(node);
    break;
  default:
}
```

the other thing I still do not get is how can I return the gathered content from inside `myCustomPlugin`? I have this currently:
```
const myCustomPlugin = (node: any) => {
  const data = node.data || (node.data = {});
  data.hProperties = node.attributes || {};
  const newNode = node;

  const markdownSource = `
# heading

* list
* items
`;

  newNode.children.unshift({
    type: "containerDirective",
    children: [
      {
        type: "html",
        value: "<div>reynier</div>"
      },
      {
        type: "html",
        value: markdownSource
      }
    ]
  });
};

export default myCustomPlugin;
```
here is some code you can play with https://codesandbox.io/s/youthful-rosalind-l9zj7d

**Reply by ChristianMurphy** [reply]
> you added the `node.name`

There is no `node.name` attribute added.
There is a `node.type` attribute added
```ts
node.data.hProperties = { type: node.name };
```

> I cannot access it inside of rehypeRenderDirective

`properties` are not accessible because you reset them before logging or switching

```ts
node.properties = {};

console.log(node.properties)
```

> the other thing I still do not get is how can I return the gathered content from inside myCustomPlugin?

A few things to consider:
1. plugins still need to be run, you have not run your plugin anywhere in the example you shared
2. don't reset properties if you want to read from them
3. make sure the label/name of the property matches what you set in the remark plugin
4. the plugin that renders is an HTML/HAST plugin, anything you pass to it need to be in `hast` format, not raw HTML or markdown, for example to go from markdown to `hast` use a snippet like the following:
   ```ts
   node.children = unified()
        .use(remarkRehype)
        // ⚠️ IMPORTANT! if content is untrusted also sanitize it with `rehype-sanitize`
        .runSync(
          unified()
           .use(remarkParse)
           .parse("# example")
        )
        .children as ElementContent[];
   ```

**Reply by reypm** [reply]
After reading and following your instructions, and if I get it correct, I am still unable to render markdown content processed from within the plugin. 

This is what I have done:

- Changed the `rehypeRenderDirective` as follow:

```tsx
function rehypeRenderDirective() {
  return (tree: HastRoot) =>
    visit(tree, { type: "element", tagName: "container-directive" }, (node) => {
      node.tagName = "div";

      let data;

      switch (node.properties.type) {
        case "my-include":
          data = myCustomPlugin(node);
          break;
        default:
      }
     
      node.children = unified()
        .use(remarkRehype)
        // ⚠️ IMPORTANT! if content is untrusted also santize it with `rehype-sanitize`
        .runSync(unified()
        .use(remarkParse)
        .parse(data))
        .children as ElementContent[];
    });
}
``` 

- Load the plugin

```
<Markdown
      remarkPlugins={[
        remarkGfm,
        remarkDirective,
        [remarkFrontmatter, ["yaml", "toml"]],
        remarkBreaks,
        remarkDirectiveCustomize,
        myCustomPlugin // here =>>> 
      ]}
      rehypePlugins={[
        rehypeRenderDirective, 
        rehypeExternalLinks, 
        rehypeSlug
      ]}
    > ...
```

- Rewrite my custom plugin as:

```
import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkRehype from "remark-rehype";
import { ElementContent } from "hast";

const myCustomPlugin = (node: any) => {
  const markdownSource = `
# heading

* list
* items
`;

  node.children = unified()
    .use(remarkRehype)
    .runSync(unified().use(remarkParse).parse(markdownSource))
    .children as ElementContent[];
};

export default myCustomPlugin;
```
but I am getting this error:

> Cannot set properties of undefined (setting 'children')

Can you take a look?

**Reply by ChristianMurphy** [reply]
It looks like both `rehypeRenderDirective` and `myCustomPlugin` are trying to convert text into an Abstract Syntax Tree? And both are rewriting the same `node`?
That is likely the source of the issue.

A pattern I'd recommend, only render or rewrite the node inside `rehypeRenderDirective`, and have another function `contentLoader` which retrieves the raw Markdown/HTML from your external system.

Avoid having another `myCustomPlugin` that isn't really a `unified` plugin and does conflicting work on the AST node.

**Reply by wooorm** [reply]
Your code is all over the place. You use `data = myCustomPlugin(...)`, but `myCustomPlugin` does not return anything? So you might be setting `undefined`? Perhaps you can clean your code? Or share more about what you actually want to do here?

**Comment by reypm** [comment]
> You use `data = myCustomPlugin(...)`, but `myCustomPlugin` does not return anything? 

Doing as follows makes the example to work:

```tsx
const myCustomPlugin = (node: any) => {
  const markdownSource = `
# heading

* list
* items
`;

  return markdownSource;
};

export default myCustomPlugin;
```

> Or share more about what you actually want to do here?

Long story short: I have a bunch of custom shortcodes for example: `email-card`, `phone-card`, and `include` among others and all of them behave differently. `phone-card` renders a component and set some data, same as `email-card` but `include` will try to include a page inside another page. 

```tsx
// this will add data to the node and later using `components` it renders some HTML code 
:::::email-card{mailto:some@some.com}Click to email something
:::::

// this will fetch the content from external sources and will return plain text and/or markdown
// is a way to reuse content
:::::
include{id=some-page}
:::::

// this will add data to the node and later using `components` it renders some HTML code
:::::phone-card{phoneNumber:1111111111}Click to make a call
:::::

// this will fetch the content from external sources and will return plain-text and/or markdown
// is a way to reuse content
:::::
include{id=some-page-1}
:::::
```

the original idea with `myCustomPlugin` was to switch over the `node.name` which before all these changes contained something like: `email-card|phone-card|include` and process those shortcodes individually on it is own file. Is this better now?

**Reply by wooorm** [reply]
Did you see the examples in the readme of `remark-directive`, sounds a lot like something like that? https://github.com/remarkjs/remark-directive#example-youtube

**Reply by reypm** [reply]
@wooorm yes, I saw that example and for the simple ones like `email-card|phone-card` it works but my problem started with `include` since what I am getting from the external resource is markdown and I need to render that back to the page

**Reply by reypm** [reply]
@wooorm / @ChristianMurphy this is how I am gathering the data from external sources:

```tsx
export default function processPageInclude(node: any, productId: string) {
  const { data } = getNodeInfo(node);
  const [cookies] = useCookies(["locale"]);
  const sharedContentPath = buildUrl(data.hProperties.id, productId);

  getApiData(sharedContentPath, cookies.locale)
    .then((data) => {
      console.log(data.data.genericSection.body); // this contains Markdown 
    })
    .catch((error: any) => {
      console.log(error);
    });
}
```
and my problem is how do I process the markdown contained at: `data.data.genericSection.body` so it renders as HTML. 🤔  

> PS: apologies for all the confusion I should have started from here. If you both want I can close this - or any of you can remove it - and we can start from scratch using this as base

**Reply by wooorm** [reply]
`toHast(fromMarkdown(value))`, where you use `mdast-util-from-markdown` and `mdast-util-to-hast`

**Reply by reypm** [reply]
@wooorm do you mean this?

```tsx
import { fromMarkdown } from 'mdast-util-from-markdown'
import { toHast } from 'mdast-util-to-hast'

...

getApiData(sharedContentPath, cookies.locale)
    .then((data) => {
      toHast(fromMarkdown(data.data.genericSection.body))
    })
    .catch((error: any) => {
      console.log(error);
    });
```

I must be missing something else since just adding that does not output anything and I am confident `data.data.genericSection.body` does contains markdown content 🤔

**Reply by wooorm** [reply]
Add some console.logs. And: `getApiData` should definitely return a hast node now? How do you use it?

**Reply by reypm** [reply]
Indeed `console.log(toHast(fromMarkdown(data.data.genericSection.body)))` will log hast content. 

How I am using it:

- The plugin gets loaded in the `ReactMarkdown` component as:
```tsx
<ReactMarkdown
  remarkPlugins={[
    ...
    [shortcodePlugin, [metaData?.id, metaData?.name]],
  ]}
  rehypePlugins={[
    rehypeExternalLinks,
    rehypeSlug
  ]}
  remarkRehypeOptions={{ allowDangerousHtml: true }}
>
  {data}
</ReactMarkdown>
```

- `shortcodePlugin` contains this:

```tsx
export default  function shortcodePlugin(options: any) {
  return (tree: any) => {
    visit(tree, (node: any) => {
      switch (node.name) {
       // ... more case
        case 'page-include':
          processPageInclude(node, options[0])
          break
        default:
      }
    })
  }
}
```

- `processPageInclude` contains:

```tsx
export default function processPageInclude(node: any, productId: string) {
  const { data } = getNodeInfo(node);
  const [cookies] = useCookies(["locale"]);
  const sharedContentPath = buildUrl(data.hProperties.id, productId);

  getApiData(sharedContentPath, cookies.locale)
    .then((data) => {
      return toHast(fromMarkdown(data.data.genericSection.body))
    })
    .catch((error: any) => {
      console.log(error);
    });
}
```

**Reply by wooorm** [reply]
You use promises. Promises are currently not yet supported in `react-markdown`. If they were, you need to await them, you can’t just randomly change things afterwards.

**Reply by reypm** [reply]
@wooorm what would be your suggestion? the only way to get the data is by calling an external URL

**Reply by wooorm** [reply, marked as answer]
Promise support, from components, in React is very much WIP. For something like https://github.com/remarkjs/react-markdown/issues/680 to land, that needs to happen first.

You can do two things:

a) create your own async component. You can use the `components` here to map custom elements (say `my-directive` to your fancy async component, and do async work there, not on the AST, but at render.
b) `react-markdown` is a tiny wrapper around other projects, look at the code, investigate the `syntax-tree` org, you can use those projects yourself, in your own async wrapper. Control the caches and stuff yourself manually

**Comment by reypm** [comment]
hey @wooorm / @ChristianMurphy 

I am still unable to get this working properly :| so here is all I have done:

```tsx
  /**
   * Gives directives a tag name, so they will transfer into the HTML AST (hast)
   */
  function remarkDirectiveCustomize(options: any) {
    return (tree: MdastRoot) =>
      visit(tree, "containerDirective", (node) => {
        node.data ??= {};
        node.data.hName = "container-directive";
        node.data.hProperties = {
          type: node.name,
          attributes: node.attributes,
          product: options[0]
        };
      });
  }

  /**
   * Find the directive HTML tags and replace them with markdown content
   */
  function rehypeRenderDirective() {
    return (tree: HastRoot) =>
      visit(
        tree,
        { type: "element", tagName: "container-directive" },
        (node) => {
          node.tagName = "div";

          if (node.properties.type === "my-include") {
            node.children = unified()
              .use(remarkRehype)
              .use(rehypeSanitize)
              .runSync(
                unified().use(remarkParse).parse(processPageInclude(node))
                  .children as ElementContent[]
              );
          }
        }
      );
  }

return (
    <ReactMarkdown
      remarkPlugins={[
        remarkGfm,
        remarkDirective,
        [remarkFrontmatter, ["yaml", "toml"]],
        remarkBreaks,
        [remarkMermaid],
        shortcodePlugin,
        [remarkDirectiveCustomize, [metaData?.id]],
      ]}
      rehypePlugins={[rehypeExternalLinks, rehypeSlug, rehypeRenderDirective]}
      remarkRehypeOptions={{ allowDangerousHtml: true }}
    >
      {data}
    </ReactMarkdown>
  );

```
But I end up with this error:
> TypeError
Expected node, got ``

But let's say the Promise is not returning any data which is not the case then I did manage to create a  [runnable example](https://codesandbox.io/s/confident-thompson-72xvf7) showing the same behavior, can you take a look and see what I am doing wrong?

**Reply by wooorm** [reply]
Right, because you set `children` to invalid things...

```js
// This is a markdown (mdast) tree, a root. It is *not* `ElementContent[]`.
const parseTree = unified().use(remarkParse).parse(processPageInclude(node))

// This is a HTML (hast) root. It is *not* `ElementContent[]`.
const transformedTree = unified().use(remarkRehype).use(rehypeSanitize).runSync(parseTree)

// @ts-expect-error: This is *almost* equivalent to `ElementContent[]`.
// It could theoretically include doctypes, but it practically currently does not.
const elementChildren: ElementContent[] = transformedTree.children

node.children = elementChildren
```

**Reply by reypm** [reply]
this is a really nice way to see things better, thanks for your help @wooorm one last thing ... should not `node.children = elementChildren` add the hast `transformedTree` to the node children? 🤔 

```tsx
  function rehypeRenderDirective() {
    return (tree: HastRoot) =>
      visit(tree, {type: "element", tagName: "container-directive" }, (node) => {
        node.tagName = "div";

        if (node.properties.type === "page-include") {
          // This is a markdown (mdast) tree, a root. It is *not* `ElementContent[]`.
          const parseTree = unified().use(remarkParse).parse(processPageInclude(node))

          // This is a HTML (hast) root. It is *not* `ElementContent[]`.
          const transformedTree = unified().use(remarkRehype).use(rehypeSanitize).runSync(parseTree)

          // @ts-expect-error: This is *almost* equivalent to `ElementContent[]`.
          // It could theoretically include doctypes, but it practically currently does not.
          const elementChildren: ElementContent[] = transformedTree.children

          node.children = elementChildren

          console.log(node)
        }
      });
  }
```

but the node does not have anything in its children:

```json
{
    "type": "element",
    "tagName": "div",
    "properties": {
        "type": "page-include",
        "attributes": {
            "id": "some-page"
        },
        "product": "commerce-sdk"
    },
    "children": [],
    "position": {
        "start": {
            "line": 1,
            "column": 1,
            "offset": 0
        },
        "end": {
            "line": 2,
            "column": 6,
            "offset": 52
        }
    }
}
```
why?

runnable example: https://codesandbox.io/s/delicate-monad-n3y45x

**Reply by wooorm** [reply]
Likely because `processPageInclude` returns `undefined`?

**Reply by reypm** [reply]
the example on codesandbox does not use Promises just plain markdown and I am having the same issue

**Reply by wooorm** [reply]
Your doing work twice. you should do work once.
`processPageInclude` already returns a hast root. You don’t need to parse that as markdown and transform that to hast. It is already hast.

```js
function processPageInclude() {
  return toHast(fromMarkdown(md))
}
```

```js
const transformedTree = processPageInclude();

// @ts-expect-error: This is *almost* equivalent to `ElementContent[]`.
// It could theoretically include doctypes, but it practically currently does not.
const elementChildren: ElementContent[] = transformedTree.children

node.children = elementChildren
```

**Reply by reypm** [reply]
thank you sir, that was the problem!
